### PR TITLE
Parse XML body on WorkflowsController#update requests

### LINKED ISSUE #88: `WorkflowsController#update` not parsing XML
The ETD robots send workflow updates via HTTP PUT, sending along an XML payload which it expects the server to parse. For example, the robot issues an HTTP PUT to `https://sul-lyberservices-test.stanford.edu/workflow/dor/objects/druid:pk473sh2989/workflows/etdSubmitWF/submit-marc`, sending the following payload:

```xml
<?xml version="1.0"?>
<process name="submit-marc" status="error" errorMessage="can't modify frozen String" errorText="sul-robots1-test.stanford.edu"/>
```

With headers of `{:content_type=>"application/xml"}`.

Instead of parsing that XML and storing the relevant bits in the database, `WorkflowsController#update` sees no `'current-status'` param and wrongly sets the workflow step as `completed`.

### PR DESCRIPTION
**Pull request by mjgiarlo**
This changes the behavior of the `WorkflowsController#update` method to be closer to that of the Java workflow service.

Fixes #88

### CODE REVIEW

File: app/controllers/workflows_controller.rb [hunk 1]
```diff
@@ -35,11 +35,14 @@ def show
...
-      find_step_for_process.update(status: params['current-status'])
-    else
-      find_step_for_process.update(status: 'completed')
-    end
+    parser = ProcessParser.new(process_from_request_body)
+    step = find_step_for_process
+
+    return render plain: process_mismatch_error(parser), status: :bad_request if parser.process != params[:process]
+
+    return render plain: status_mismatch_error(step), status: :precondition_failed if params['current-status'] && step.status != params['current-status']
```
Comment (jcoyne) [inline app/controllers/workflows_controller.rb]: This is probably the wrong status code. It doesn’t match the spec https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412

File: app/services/workflow_parser.rb [hunk 2]
```diff
@@ -28,13 +28,19 @@ def create_workflow_steps
     end
   end
```
Comment (jcoyne) [inline app/services/workflow_parser.rb]: Why does this become public?

File: app/controllers/workflows_controller.rb (hunk 1 above)
Comment (mjgiarlo) [reply app/controllers/workflows_controller.rb]: Good call.

File: app/services/workflow_parser.rb (hunk 2 above)
Comment (mjgiarlo) [reply app/services/workflow_parser.rb]: From a prior tweak that I backed away from. Will revert. Good eye!

File: app/controllers/workflows_controller.rb (hunk 1 above)
Comment (mjgiarlo) [reply app/controllers/workflows_controller.rb]: @jcoyne How about 409 Conflict? 400 Bad Request? other?

**Comment by mjgiarlo** [comment]
@jcoyne 	Changes made and pushed.

File: app/controllers/workflows_controller.rb [hunk 3]
```diff
@@ -35,11 +35,14 @@ def show
   end
 
   def update
-    if params['current-status'].present?
-      find_step_for_process.update(status: params['current-status'])
-    else
-      find_step_for_process.update(status: 'completed')
-    end
+    parser = ProcessParser.new(process_from_request_body)
```
Comment (mjgiarlo) [inline app/controllers/workflows_controller.rb]: I don't love the shape of this method. Happy to hear ideas for how to clean it up or refactor further.

File: app/services/process_parser.rb [hunk 4]
```diff
@@ -3,27 +3,65 @@
...
+      next if value.nil?
+
+      hash[attribute] = value
+    end
+    hash
   end
 
-  def name
-    process.attr('name')
+  def process
```
Comment (mjgiarlo) [inline app/services/process_parser.rb]: Renamed this method in order to align the names of methods in this class with the names of fields in the database. This is what allows us to call `step.update(parser.to_h)` in the workflows controller.

File: app/services/process_parser.rb [hunk 5]
```diff
@@ -3,27 +3,65 @@
 ##
 # Parsing Process creation request
 class ProcessParser
-  attr_reader :process
+  attr_reader :process_element
+
+  PROCESS_ATTRIBUTES = %i[
```
Comment (mjgiarlo) [inline app/services/process_parser.rb]: I believe this is the full list of attributes from the Java-based workflow service.

File: spec/factories/workflow_steps.rb [hunk 6]
```diff
@@ -6,7 +6,7 @@
       "druid:abc123#{n}"
     end
     datastream { 'accessionWF' }
-    process { 'proc' }
+    process { 'start-accession' }
```
Comment (mjgiarlo) [inline spec/factories/workflow_steps.rb]: Line up the factory with the XML fixture data.

File: app/controllers/workflows_controller.rb [hunk 7]
```diff
@@ -35,11 +35,14 @@ def show
...
+    step = find_step_for_process
+
+    return render plain: process_mismatch_error(parser), status: :bad_request if parser.process != params[:process]
+
+    return render plain: status_mismatch_error(step), status: :conflict if params['current-status'] && step.status != params['current-status']
+
+    step.update(parser.to_h)
```
Comment (mjgiarlo) [inline app/controllers/workflows_controller.rb]: The prior implementation allowed PUT requests containing no XML body, which had the effect of setting the status to `completed`. Can we verify that use case? It no longer exists after this PR; we now require an XML payload to set workflow step attributes.

File: app/controllers/workflows_controller.rb (hunk 1 above)
Comment (mjgiarlo) [reply app/controllers/workflows_controller.rb]: Went with 409 due to conflict in current status.

**Pull request merged**